docs(headless): typedoc website and annotations

jpmarceau opened the pull request:
[KIT-3422](https://coveord.atlassian.net/browse/KIT-3422)

Follows https://github.com/coveo/ui-kit/pull/4361, which prepared the code to build a typedoc website. I couldn't merge it because the incomplete tagging state was causing build problems. I carry over the code changes from that PR in this one here, but they've already been reviewed. We should be able to merge them with this PR here.

Now, the main goal of this PR is adding lots of tags to organize the contents of the Typedoc website. That's why there are lots of file changes; they're very simple and repetitive.
(I also added missing documentation here and there, as I stumbled upon it)

Rather than looking through all files, I recommend looking through the built site and letting me know what you think. Before I keep adding/tweaking things, now would be a good time for feedback 🙂 

To look through the built site, either:

1: unzip the following and open `index.html`:
[docs.zip](https://github.com/user-attachments/files/17497151/docs.zip)


or

2: 
    - check out my branch
    - run `npm install`
    - run `npm run build:typedoc` from the headless package
    - open `packages/headless/docs/index.html`

[KIT-3422]: https://coveord.atlassian.net/browse/KIT-3422?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

alexprudhomme commented on packages/headless/doc-parser/package-lock.json:
```diff
@@ -0,0 +1,325 @@
+{
+  "name": "doc-parser",
```
As a solution, I think you have to delete them both and then do `npm install` from the root.

jpmarceau replied:
Good catch
I must've run "npm install" from the headless package rather than the root
Fixed in next commit

alexprudhomme commented on packages/headless/src/commerce.index.ts:
````diff
@@ -1,3 +1,20 @@
+ *
+ * @example
+ * ```typescript
+ * import { buildCommerceEngine, getSampleCommerceEngineConfiguration, buildSearch } from '@coveo/headless-commerce';
````
```suggestion
 * import { buildCommerceEngine, getSampleCommerceEngineConfiguration, buildSearch } from '@coveo/headless/commerce';
```

alexprudhomme commented on packages/headless/src/controllers/core/did-you-mean/headless-core-did-you-mean.ts:
```diff
@@ -95,6 +110,9 @@ export interface DidYouMeanState {
  * @param engine - The headless engine.
  * @param props - The configurable `DidYouMean` controller properties.
  * @returns A `DidYouMean` controller instance.
+ *
+ * @group Core controllers
+ * @category DidYouMean
  */
 export function buildCoreDidYouMean(
   engine: CoreEngine,
```
Is there a way for this controller to not be documented ? It is not exported in the index.ts files. Same with most of the stuff in core/.

jpmarceau replied:
I just realized that `buildCoreDidYouMean` isn't exported, so my tags here are superfluous. I'll need to do a pass in general to remove unnecessary tags.

I do think I need to export the state and controller interfaces though, because they're referenced by both the Search and Insight buildDidYouMean functions...

alexprudhomme commented on packages/headless/src/controllers/core/context/headless-core-context.ts:
```diff
@@ -95,6 +106,9 @@ export interface ContextState {
  * @param props - The configurable `Context` controller properties.
  *
  * @returns a `Context` controller instance.
+ *
+ * @group Core controllers
+ * @category Context
  */
 export function buildCoreContext(
```
This one is weird as well it is only used in headless-context and in headless-strict-context-factory. @louis-bompart @y-lakhdar, I saw you two worked on those controllers. Can you give more context ?

It should not be in core/ but in search/

jpmarceau replied:
That kind of thing happens in a lot of controllers, that's why I include a lot of "Core controllers". For a lot of Search components, the "buildX" method is exposed as a normal "Search controller", but the state and interface are in a different, /core file.
For example, `buildCategoryFacet` is in headless-category-facets.ts, 
<img width="389" alt="Screenshot 2024-10-11 at 3 16 46 PM" src="https://github.com/user-attachments/assets/a24936c4-6d67-486b-b140-d8decd4d1cd6">
but the state and interface are in headless-core-category-facets.ts
<img width="1332" alt="Screenshot 2024-10-11 at 3 19 00 PM" src="https://github.com/user-attachments/assets/bf1191ab-d885-451d-b856-07203923e97b">

jpmarceau replied:
Furthermore, the core controllers states and interfaces can be referenced from builder functions accessible through different entry points. For instance, here the Insight CategoryFacet links to the core category facet interface and state
<img width="1318" alt="Screenshot 2024-10-11 at 3 24 30 PM" src="https://github.com/user-attachments/assets/b1604cb0-2523-4a4e-91c0-7647938e4915">

alexprudhomme reviewed (commented):
Would it be possible to have all `core/` stuff in another typedoc module or just not there ?

jpmarceau commented:
> Would it be possible to have all `core/` stuff in another typedoc module or just not there ?

Good question
I think we have to keep them somehow. As commented earlier, core states and interfaces can be referenced by multiple sub-packages.
I haven't found a way to create a "Core" module to group them so far, but I'll keep looking :loading:

jpmarceau commented:
Update: I've dug more to answer your question @alexprudhomme , and I think we have limited options.
Something to know: [TypeDoc modules are created based on the entry points we give it](https://github.com/TypeStrong/typedoc/issues/2745#issuecomment-2407911986), with basically no customization. This means that if we wanted to separate out the Core controller stuff in a separate module (alongside Search, Case Assist, etc.), we would need to create a separate `core.index.ts` file specifically to export and document separately the Core controller stuff.
I'm not sure we'd want to do that, but I've tried it anyway, just to see. It turns out it's not really useful, because the vast majority of the Core controller stuff is still exported under as types from other sub-packages anyway. E.g.:
<img width="691" alt="Screenshot 2024-10-14 at 1 01 47 PM" src="https://github.com/user-attachments/assets/52eed124-b70f-47f6-82f0-2c67b131992f">
So, such core controllers always still show up under the Search module, or whichever other modules re-export them...

I think we have two options:
1- Adding the `@internal` tag on the Core stuff we don't want to document publicly. However, I doubt that we'd want to do that very often, because most of the Core stuff that's currently documented is referenced by builder functions in the Search and Insight subpackages, and sometimes others... I think it's typically useful.

2- I could drop the `Core` from `Core controllers`, that way all the useful `Core controllers` would show up within the `Search` module under `Controller`, alongside the builder functions which are currently there. For example, for the CategoryFacet stuff:
Before:

![Screenshot 2024-10-14 at 1 11 22 PM](https://github.com/user-attachments/assets/bbc5d959-94a3-41f3-a074-62e25cae1eab)
![Screenshot 2024-10-14 at 1 11 27 PM](https://github.com/user-attachments/assets/abac541a-a832-4ea6-900a-6fda86f5ec34)
![Screenshot 2024-10-14 at 1 11 36 PM](https://github.com/user-attachments/assets/007ebdc9-281f-497b-bbbb-ee58f6451918)

After:

![Screenshot 2024-10-14 at 1 12 47 PM](https://github.com/user-attachments/assets/c483d400-b540-4fe1-b58a-5b52c8621c6e)
![Screenshot 2024-10-14 at 1 12 54 PM](https://github.com/user-attachments/assets/ce143706-38c9-47bc-bdfd-2ce08d8df8f3)

Personally, I don't find that as clear. I think it's better to have the core controllers as I currently have them, because it's a bit clearer now why they can be used by several modules.

Gerrit0 commented:
> So, such core controllers always still show up under the Search module, or whichever other modules re-export them...

TypeDoc has an [excludeReferences](https://typedoc.org/options/input/#excludereferences) option. If specified, and `core` is listed first in the entry points, they should only show up under that module.

jpmarceau commented:
> > So, such core controllers always still show up under the Search module, or whichever other modules re-export them...
> 
> TypeDoc has an [excludeReferences](https://typedoc.org/options/input/#excludereferences) option. If specified, and `core` is listed first in the entry points, they should only show up under that module.

Thanks @Gerrit0, that did it 👌

jpmarceau commented:
Update for reviewers:

Following @alexprudhomme and @Gerrit0 's suggestion, I added a [core.index.ts](https://github.com/coveo/ui-kit/pull/4517/files#diff-acc1b2d2410736aa63b20194af5b6eb103363b9289312ed355cf91a230b83053) file that exports the core stuff, so that Typedoc can put it under its own module. In other words, the core documentation now shows up under "Core" rather than anything else.
![Screenshot 2024-10-23 at 3 28 04 PM](https://github.com/user-attachments/assets/dd3f983d-da12-440e-97b9-8e81738efe89)
(The core.index.ts file is only used an an entry point for typedoc. It shouldn't have any impact beyond that on the headless project.)

I also added a new [plugin](https://github.com/coveo/ui-kit/pull/4517/files#diff-4c3ce285debcc6c0ef47db2fec60d9327a95faab2ce383c1cbc718f531ad9281) to easily add repetitive comments across many files.

Ready for review ✅

alexprudhomme commented on packages/headless/typedoc.json:
```diff
@@ -0,0 +1,30 @@
+    "./temp/index.json",
+    "./temp/ssr.json",
+    "./temp/commerce.json",
+    // "./temp/ssr-commerce.json", //TODO - KIT-3690 : Uncomment when the SSR Commerce is ready for GA
```
uncomment this at GA here https://coveord.atlassian.net/browse/KIT-3690

fbeaudoincoveo reviewed (approved):
Awesome

alexprudhomme commented on packages/headless/package.json:
```diff
@@ -136,6 +136,9 @@
+    "build:typedoc:docs": "typedoc --tsconfig tsconfig.typedoc.json",
     "build:doc": "npm run build:doc:extract && npm run build:doc:parse",
     "build:doc:extract": "node ./scripts/extract-documentation.mjs",
     "build:doc:parse": "ts-node --project ./doc-parser/tsconfig.build.json ./doc-parser/doc-parser.ts"
```
does not fail now
